src/Instrumentation/Doctrine/src/DoctrineInstrumentation.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\Doctrine;
+
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\API\Trace\SpanBuilderInterface;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\API\Trace\StatusCode;
+use OpenTelemetry\Context\Context;
+
+use function OpenTelemetry\Instrumentation\hook;
+
+use OpenTelemetry\SemConv\TraceAttributes;
+use Throwable;
+
+class DoctrineInstrumentation
+{
+    public const NAME = 'doctrine';
+
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.doctrine');
+
+        hook(
+            \Doctrine\DBAL\Driver::class,
+            'connect',
+            pre: static function (\Doctrine\DBAL\Driver $driver, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'Doctrine\DBAL\Driver::connect', $function, $class, $filename, $lineno)
+                    ->setSpanKind(SpanKind::KIND_CLIENT);
```

Comment:
How do we think about the duality of PDO and dbal here? Most instrumented methods have a pendant in the PDO instrumentation i.e. - a compatible pdo driver will cause two client spans and the hierarchy will probably be

```
my_func (internal)
 -> connect (client) # from dbal
   -> PDO::__construct (client ) # from pdo
```

IMO the two should be either mutually exclusive or enrich each other. WDYT?

Reply:
We do not have a way for two auto-instrumentations to cooperate like this (it's an idea that has come up before, and would be cool). So I think mutually exclusive is the best we have - or at least it's up to the user to install only one instrumentation, I don't think we need to go as far as setting them up as conflicting in composer.
Is documenting that recommendation good enough for now?

Reply:
Could the PDO instrumentation re-write the doctrine span from CLIENT to internal or bail if it recognizes the parent is already client?

Reply:
To your point - i think documenting it makes the experience worse because users then need to make a whole lot of decisions when instrumenting

Reply:
> i think documenting it makes the experience worse

I only meant documenting such as "use pdo or doctrine auto-instrumentation, but probably not both" - I expect users to make decisions about which auto-instrumentations to install based on their workload/stack...just installing everything available is probably too noisy.

> Could the PDO instrumentation re-write the doctrine span from CLIENT to internal or bail if it recognizes the parent is already client?

I think the issue here would be that pre/post hooks operate in isolation, so there's currently no way for a post hook to know whether the pre hook created a span or just modified the active span. All the implementations we have assume that `pre` created and activated a span, and `post` just closes whichever is the active span at the time.
We have previously brainstormed whether we could manage some state between a pre and post hook, but we thought it was going to be a hard problem.

Reply:
> I expect users to make decisions about which auto-instrumentations to install based on their workload/stack...just installing everything available is probably too noisy.

I completely agree on this matter.

Reply:
> Could the PDO instrumentation ... bail if it recognizes the parent is already client?

Bailing out requires storing additional data within the `Context` to be able to detect duplicate instrumentations, see Javas span suppression strategies:
[Instrumentation span suppression behavior](https://opentelemetry.io/docs/zero-code/java/agent/disable/#instrumentation-span-suppression-behavior)
[SpanSuppressionStrategy.java](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressionStrategy.java)
[SpanSuppressors.java](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java)

> AFAIK the idea is to allow the agent to do that

Dropping at the agent is too late as it may lead to broken traces "Dropping a span may lead to orphaned spans if the dropped span is a parent." [ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#warnings).

> there's currently no way for a post hook to know whether the pre hook created a span or just modified the active span

Scopes attached via `Context::storage()->attach()` implement `ArrayAccess` to allow propagating state from `pre` hook to `post` hook [ref](https://github.com/open-telemetry/opentelemetry-php/pull/675).

```php
hook(
    null,
    'example',
    static function() use ($tracer): void {
        $context = Context::getCurrent();
        if (lcg_value() > .5 /* some suppression condition */) {
            $span = $tracer
                ->spanBuilder('example')
                ->startSpan();
            $context = $span->storeInContext($context);
        }
        $scope = Context::storage()->attach($context);
        $scope[SpanInterface::class] = $span ?? null;
    },
    static function(): void {
        if (!$scope = Context::storage()->scope()) {
            return;
        }
        $scope->detach();
        /** @var SpanInterface|null $span */
        $span = $scope[SpanInterface::class] ?? null;
        $span?->end();
    },
);
```